lib/clean/invalid_subscriber_lists.rb

```diff
@@ -0,0 +1,146 @@
+module Clean
+  # InvalidSubscriberLists enables us to find and correct subscriber lists
+  # with invalid tags.
+  class InvalidSubscriberLists
+    def lists
+      SubscriberList.select(&:invalid_tags?)
+    end
+
+    def valid_list(invalid_list, dry_run: true)
+      return unless subscriber_lists?([invalid_list])
+
+      unless invalid_list.invalid_tags?
+        raise StandardError, 'List is not invalid'
+      end
+
+      slug = invalid_list.slug + '-untagged'
+      list = SubscriberList.find_by(slug: slug)
+      return list unless list.nil?
+
+      new_list = SubscriberList.new(
+        document_type: invalid_list.document_type,
+        email_document_supertype: invalid_list.email_document_supertype,
+        government_document_supertype: invalid_list.government_document_supertype,
+        links: tags_to_links(invalid_list.tags),
+        slug: slug,
+        title: invalid_list.title,
+      )
+
+      new_list.save! unless dry_run
+
+      new_list
+    rescue GdsApi::ContentStore::ItemNotFound => e
+      puts "InvalidTagValueError: Did not create subscribers of list #{invalid_list.slug}
+        as we couldn't create a valid subscriber list: #{e.message}"
+    end
+
+    def copy_subscribers(from_list, to_list, dry_run: true)
+      return unless subscriber_lists?([from_list, to_list])
+
+      subscribers = from_list.subscribers.activated
+      total_subscribers = subscribers.count
+
+      subscribers.each do |subscriber|
+        Subscription.transaction do
+          existing_subscription = Subscription.active.find_by(
+            subscriber: subscriber,
+            subscriber_list: from_list,
+          )
+
+          next unless existing_subscription
+
+          # Check if they have already subscribed
+          subscribed_to_destination_subscriber_list = Subscription.find_by(
+            subscriber: subscriber,
+            subscriber_list: to_list
+          )
+
+          # Subscribed them if they haven't already been subscribed
+          if subscribed_to_destination_subscriber_list.nil?
+            unless dry_run
+              Subscription.create!(
```

Comment:
Just a question about using the `!` ... Is it impossible for the validations to fail? Because there's no rescue if it does.

Reply:
The validations could fail, but that is OK. I suspect they won't, as the data is already in the DB, the task is just putting it back in a new list. I would prefer the task to fail if validations do fail, and we can always safely restart the task if that happens.